torch/distributed/hooks.py

```diff
@@ -0,0 +1,169 @@
+import logging
+import os
+import threading
+import time
+from dataclasses import dataclass
+from typing import Callable, Dict, List, Optional
+
+import torch.distributed as dist
+import torch.distributed.distributed_c10d as c10d
+
+from torch._C._distributed_c10d import (
+    _dequeue_c10d_event,
+    _enable_event_collection,
+    EventKind,
+)
+
+__all__ = [
+    "CollectiveStatus",
+    "COLLECTIVE_HOOK_TYPE",
+    "PG_HOOK_TYPE",
+    "register_collective_start_hook",
+    "register_collective_end_hook",
+    "register_process_group_hook",
+]
+
+
+@dataclass
+class CollectiveStatus:
+    pg_name: str = "unknown"  # This name matches the one informed in the pgreg cb
+    backend: str = "unknown"  # Name of the backend used
+    sequence_number: int = -1  # This name matches the one informed in the pgreg cb
+    operation: str = "unknown"  # collective name
+    timestamp: int = 0  # timestamp to the earliest time we noticed this event
+    duration: Optional[float] = None  # value in milliseconds it took executing
+    drop_count: int = 0  # number of events dropped following this one
+
+
+COLLECTIVE_HOOK_TYPE = Callable[[CollectiveStatus], None]
+PG_HOOK_TYPE = Callable[[dist.ProcessGroup, str], None]
+
+# This controls the number of internal failures we'll tolerate before giving up
+_MAX_INTERNAL_FAILURES = 10
+
+logger = logging.getLogger(__name__)
+_cb_thread: Optional[threading.Thread] = None
+_start_callbacks: List[COLLECTIVE_HOOK_TYPE] = []
+_end_callbacks: List[COLLECTIVE_HOOK_TYPE] = []
+_pp_r = -1
+_pp_w = -1
+
+
+def _c10d_pg_hooks_loops():
+    internal_failures = 0
+    while True:
```

Comment:
nit: Maybe we can add some documents explaining that we need to keep Python alive so that cpp watchdog can call the callback and do something in python. Not need to be done in this PR, though.

Reply:
well that is not really true imo.  we should not design our watchdog on top of python hooks probably.  i am not sure if we will use python hooks for production or if we will use them for experimentation.  the later PR in the stack adds c++ hooks.  we probably should use those if we build a critical component on top of them.  We could still use python hooks for some telemtry i think.